File: macos/QMK Toolbox/HIDConsole/HIDConsoleListener.m
#import "HIDConsoleListener.h"

const ushort CONSOLE_USAGE_PAGE = 0xFF31;
const ushort CONSOLE_USAGE = 0x0074;

@implementation HIDConsoleListener
- (id)init {
    if (self = [super init]) {
        self.hidManagerRef = IOHIDManagerCreate(kCFAllocatorDefault, kIOHIDOptionsTypeNone);

        CFMutableDictionaryRef consoleMatcher = [self matchingDictionaryForUsagePage:CONSOLE_USAGE_PAGE usage:CONSOLE_USAGE];
        consoleMatcher = (CFMutableDictionaryRef) CFRetain(consoleMatcher);
        IOHIDManagerSetDeviceMatching(self.hidManagerRef, consoleMatcher);
        CFRelease(consoleMatcher);
    }
    return self;
}

- (void)start {
    if (self.devices == nil) {
        self.devices = [[NSMutableArray alloc] init];
    }

    IOHIDManagerScheduleWithRunLoop(self.hidManagerRef, CFRunLoopGetCurrent(), kCFRunLoopDefaultMode);
    IOHIDManagerOpen(self.hidManagerRef, kIOHIDOptionsTypeNone);

    IOHIDManagerRegisterDeviceMatchingCallback(self.hidManagerRef, deviceConnected, (__bridge void *)self);
    IOHIDManagerRegisterDeviceRemovalCallback(self.hidManagerRef, deviceDisconnected, (__bridge void *)self);
}

- (CFMutableDictionaryRef)matchingDictionaryForUsagePage:(ushort)usagePage usage:(ushort)usage {
    CFMutableDictionaryRef result = CFDictionaryCreateMutable(kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);

    CFNumberRef cfUsagePage = CFNumberCreate(kCFAllocatorDefault, kCFNumberShortType, &usagePage);
    CFDictionarySetValue(result, CFSTR(kIOHIDDeviceUsagePageKey), cfUsagePage);
    CFRelease(cfUsagePage);

    CFNumberRef cfUsage = CFNumberCreate(kCFAllocatorDefault, kCFNumberShortType, &usage);
    CFDictionarySetValue(result, CFSTR(kIOHIDDeviceUsageKey), cfUsage);
    CFRelease(cfUsage);

    return result;
}

static void deviceConnected(void *context, IOReturn result, void *sender, IOHIDDeviceRef device) {
    HIDConsoleListener *const listener = (__bridge HIDConsoleListener *const)context;

    BOOL alreadyListed = NO;
    for (HIDConsoleDevice *d in listener.devices) {
        if (d.deviceRef == device) {
            alreadyListed = YES;
            break;
        }
    }
    if (!alreadyListed) {
        HIDConsoleDevice *consoleDevice = [[HIDConsoleDevice alloc] initWithDeviceRef:device];
        consoleDevice.delegate = listener;
        [listener.devices addObject:consoleDevice];
        [listener.delegate consoleDeviceDidConnect:consoleDevice];
    }
}

static void deviceDisconnected(void *context, IOReturn result, void *sender, IOHIDDeviceRef device) {
    HIDConsoleListener *const listener = (__bridge HIDConsoleListener *const)context;

    for (HIDConsoleDevice *d in listener.devices) {
        if (d.deviceRef == device) {
            HIDConsoleDevice *tempDevice = [d copy];
            [listener.devices removeObject:d];
            [listener.delegate consoleDeviceDidDisconnect:tempDevice];
            break;
        }
    }
}

-(void)consoleDevice:(HIDConsoleDevice *)device didReceiveReport:(NSString *)report {
    [self.delegate consoleDevice:device didReceiveReport:report];
}

- (void)stop {
    IOHIDManagerRegisterDeviceMatchingCallback(self.hidManagerRef, NULL, (__bridge void *)self);
    IOHIDManagerRegisterDeviceRemovalCallback(self.hidManagerRef, NULL, (__bridge void *)self);
    IOHIDManagerUnscheduleFromRunLoop(self.hidManagerRef, CFRunLoopGetCurrent(), kCFRunLoopDefaultMode);
    IOHIDManagerClose(self.hidManagerRef, kIOHIDManagerOptionNone);
}
@end

File: macos/QMK Toolbox/HIDConsole/HIDConsoleListener.swift
import Foundation
import IOKit.hid

let CONSOLE_USAGE_PAGE: UInt16 = 0xFF31
let CONSOLE_USAGE: UInt16      = 0x0074

protocol HIDConsoleListenerDelegate: AnyObject {
    func consoleDeviceDidConnect(_ device: HIDConsoleDevice)

    func consoleDeviceDidDisconnect(_ device: HIDConsoleDevice)

    func consoleDevice(_ device: HIDConsoleDevice, didReceiveReport report: String)
}

class HIDConsoleListener: HIDConsoleDeviceDelegate {
    weak var delegate: HIDConsoleListenerDelegate?

    private var hidManager: IOHIDManager

    var devices: [HIDConsoleDevice] = []

    init() {
        hidManager = IOHIDManagerCreate(kCFAllocatorDefault, IOOptionBits(kIOHIDOptionsTypeNone))
        let consoleMatcher = [kIOHIDDeviceUsagePageKey: CONSOLE_USAGE_PAGE, kIOHIDDeviceUsageKey: CONSOLE_USAGE]
        IOHIDManagerSetDeviceMatching(hidManager, consoleMatcher as CFDictionary?)
    }

    func start() {
        IOHIDManagerScheduleWithRunLoop(hidManager, RunLoop.current.getCFRunLoop(), CFRunLoopMode.defaultMode.rawValue)
        IOHIDManagerOpen(hidManager, IOOptionBits(kIOHIDOptionsTypeNone))

        let matchingCallback: IOHIDDeviceCallback = { context, result, sender, device in
            let listener: HIDConsoleListener = Unmanaged<HIDConsoleListener>.fromOpaque(context!).takeUnretainedValue()
            listener.deviceConnected(device)
        }

        let removalCallback: IOHIDDeviceCallback = { context, result, sender, device in
            let listener: HIDConsoleListener = Unmanaged<HIDConsoleListener>.fromOpaque(context!).takeUnretainedValue()
            listener.deviceDisconnected(device)
        }

        let unsafeSelf = Unmanaged.passRetained(self).toOpaque()
        IOHIDManagerRegisterDeviceMatchingCallback(hidManager, matchingCallback, unsafeSelf)
        IOHIDManagerRegisterDeviceRemovalCallback(hidManager, removalCallback, unsafeSelf)
    }

    func deviceConnected(_ device: IOHIDDevice) {
        if devices.contains(where: { $0.hidDevice === device }) {
            return
        }

        let consoleDevice = HIDConsoleDevice(device)
        consoleDevice.delegate = self
        devices.append(consoleDevice)
        delegate?.consoleDeviceDidConnect(consoleDevice)
    }

    func deviceDisconnected(_ device: IOHIDDevice) {
        let discardedItems = devices.filter { $0.hidDevice === device }
        devices = devices.filter { !discardedItems.contains($0) }

        for d in discardedItems {
            delegate?.consoleDeviceDidDisconnect(d)
        }
    }

    func consoleDevice(_ device: HIDConsoleDevice, didReceiveReport report: String) {
        delegate?.consoleDevice(device, didReceiveReport: report)
    }

    func stop() {
        let unsafeSelf = Unmanaged.passRetained(self).toOpaque()
        IOHIDManagerRegisterDeviceMatchingCallback(hidManager, nil, unsafeSelf)
        IOHIDManagerRegisterDeviceRemovalCallback(hidManager, nil, unsafeSelf)
        IOHIDManagerUnscheduleFromRunLoop(hidManager, CFRunLoopGetCurrent(), CFRunLoopMode.defaultMode.rawValue)
        IOHIDManagerClose(hidManager, IOOptionBits(kIOHIDOptionsTypeNone))
    }
}
